src/algebra/direct_sum/internal.lean

```diff
@@ -137,28 +137,150 @@ instance gcomm_ring [add_comm_monoid ι] [comm_ring R] [set_like σ R]
-  ((r * r') i : R) =
-    ∑ ij in (r.support ×ˢ r'.support).filter (λ ij : ι × ι, ij.1 + ij.2 = i), r ij.1 * r' ij.2 :=
+lemma direct_sum.coe_mul_apply [add_monoid ι] [set_like.graded_monoid A]
+  [Π (i : ι) (x : A i), decidable (x ≠ 0)] (r r' : ⨁ i, A i) (n : ι) :
+  ((r * r') n : R) =
+    ∑ ij in (r.support ×ˢ r'.support).filter (λ ij : ι × ι, ij.1 + ij.2 = n), r ij.1 * r' ij.2 :=
 begin
   rw [direct_sum.mul_eq_sum_support_ghas_mul, dfinsupp.finset_sum_apply,
     add_submonoid_class.coe_finset_sum],
   simp_rw [direct_sum.coe_of_apply, ←finset.sum_filter, set_like.coe_ghas_mul],
 end
 
+lemma direct_sum.coe_mul_apply_eq_dfinsupp_sum [add_monoid ι] [set_like.graded_monoid A]
+  [Π (i : ι) (x : A i), decidable (x ≠ 0)] (r r' : ⨁ i, A i) (n : ι) :
+  ((r * r') n : R) = r.sum (λ i ri, r'.sum (λ j rj, if i + j = n then ri * rj else 0)) :=
+begin
+  simp only [direct_sum.mul_eq_dfinsupp_sum, dfinsupp.sum_apply],
+  iterate 2 { rw [dfinsupp.sum, add_submonoid_class.coe_finset_sum], congr, ext },
+  dsimp only, split_ifs,
+  { subst h, rw direct_sum.of_eq_same, refl },
+  { rw direct_sum.of_eq_of_ne _ _ _ _ h, refl },
+end
+
+lemma direct_sum.coe_of_mul_apply_aux [add_monoid ι] [set_like.graded_monoid A] {i : ι}
+  (r : A i) (r' : ⨁ i, A i) {j n : ι} (H : ∀ (x : ι), i + x = n ↔ x = j) :
+  ((direct_sum.of _ i r * r') n : R) = r * r' j :=
+begin
+  classical,
+  rw direct_sum.coe_mul_apply_eq_dfinsupp_sum,
+  apply (dfinsupp.sum_single_index _).trans, swap,
+  { simp_rw [add_submonoid_class.coe_zero, zero_mul, if_t_t], exact dfinsupp.sum_zero },
+  simp_rw [dfinsupp.sum, H, finset.sum_ite_eq'],
+  split_ifs, refl,
+  rw [dfinsupp.not_mem_support_iff.mp h, add_submonoid_class.coe_zero, mul_zero],
+end
+
+lemma direct_sum.coe_mul_of_apply_aux [add_monoid ι] [set_like.graded_monoid A]
+  (r : ⨁ i, A i) {i : ι} (r' : A i) {j n : ι} (H : ∀ (x : ι), x + i = n ↔ x = j) :
+  ((r * direct_sum.of _ i r') n : R) = r j * r' :=
+begin
+  classical,
+  rw [direct_sum.coe_mul_apply_eq_dfinsupp_sum, dfinsupp.sum_comm],
+  apply (dfinsupp.sum_single_index _).trans, swap,
+  { simp_rw [add_submonoid_class.coe_zero, mul_zero, if_t_t], exact dfinsupp.sum_zero },
+  simp_rw [dfinsupp.sum, H, finset.sum_ite_eq'],
+  split_ifs, refl,
+  rw [dfinsupp.not_mem_support_iff.mp h, add_submonoid_class.coe_zero, zero_mul],
+end
+
+lemma direct_sum.coe_of_mul_apply_eq_ite [add_left_cancel_monoid ι] [set_like.graded_monoid A]
+  {i : ι} (r : A i) (r' : ⨁ i, A i) (n : ι) [decidable (∃ j, i + j = n)] :
+  ((direct_sum.of _ i r * r') n : R) = if h : ∃ j, i + j = n then r * r' h.some else 0 :=
+begin
+  classical, split_ifs,
+  { exact direct_sum.coe_of_mul_apply_aux _ _ _
+     (λ j, ⟨λ he, add_left_cancel (he.trans h.some_spec.symm), λ he, he.symm ▸ h.some_spec⟩) },
+  rw direct_sum.coe_mul_apply_eq_dfinsupp_sum,
+  apply (dfinsupp.sum_single_index _).trans, swap,
+  { simp_rw [add_submonoid_class.coe_zero, zero_mul, if_t_t], exact dfinsupp.sum_zero },
```

Comment:
This is what I don't like: large amount of repetition between proofs. Adopting [this formulation](https://github.com/leanprover-community/mathlib/pull/16083#issuecomment-1219100230) could reduce the duplication. What do you think, @eric-wieser ?

Reply:
Oh never mind, that formulation is actually wrong (again)! Sorry for the disturbance.